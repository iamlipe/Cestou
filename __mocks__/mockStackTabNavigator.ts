export const mockStackTabNavigatorConsumer = {
  key: '',
  index: 0,
  routeNames: [],
  history: [],
  routes: [
    {
      key: '',
      name: 'HomeConsumer',
      params: undefined,
    },
    {
      key: '',
      name: 'DonationConsumer',
      params: undefined,
    },
    {
      key: '',
      name: 'ProfileConsumer',
      params: undefined,
    },
    {
      key: '',
      name: 'BasketConsumer',
      params: undefined,
    },
  ],
  stale: false,
  type: 'tab',
};

export const mockStackTabNavigatorProducer = {
  history: [],
  index: 0,
  key: '',
  routeNames: ['HomeProducer', 'FinancialProducer', 'ProfileProducer'],
  routes: [
    {
      key: '',
      name: 'HomeProducer',
      params: undefined,
    },
    {
      key: '',
      name: 'FinancialProducer',
      params: undefined,
    },
    {
      key: '',
      name: 'ProfileProducer',
      params: undefined,
    },
  ],
  stale: false,
  type: 'tab',
};
