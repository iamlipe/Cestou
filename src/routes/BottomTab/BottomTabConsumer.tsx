import React from 'react';
import styled, {useTheme} from 'styled-components/native';
import {BottomTabBarProps} from '@react-navigation/bottom-tabs';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {LoggedConsumerStackParamList} from '../stacks/LoggedConsumerStack';

// icons
import TabHomeIcon from '@/assets/svgs/tab-home.svg';
import TabBasketIcon from '@/assets/svgs/tab-basket.svg';
import TabFinacial from '@/assets/svgs/tab-financial.svg';
import TabProfileIcon from '@/assets/svgs/tab-profile.svg';

type NavPropsProducer = NativeStackNavigationProp<
  LoggedConsumerStackParamList,
  'HomeConsumer' | 'BasketConsumer' | 'DonationConsumer' | 'ProfileConsumer'
>;

export const ButtonTabConsumer: React.FC<BottomTabBarProps> = ({state}) => {
  const {navigate} = useNavigation<NavPropsProducer>();
  const activeTab = state.routes[state.index].name;
  const theme = useTheme();

  const renderTab = (
    icon: JSX.Element,
    title: string,
    route: keyof LoggedConsumerStackParamList,
  ) => (
    <StyledButtonInactive onPress={() => navigate(route)}>
      {icon}
      {activeTab === route ? (
        <StyledTextActive>{title}</StyledTextActive>
      ) : (
        <StyledTextInactive>{title}</StyledTextInactive>
      )}
    </StyledButtonInactive>
  );

  return (
    <StyledBottomTabContainer>
      <StyledButtonTabRow>
        {renderTab(
          <TabHomeIcon
            fill={
              activeTab === 'HomeConsumer'
                ? theme.colors.PRIMARY_600
                : theme.colors.GRAY_700
            }
          />,
          'Início',
          'HomeConsumer',
        )}

        {renderTab(
          <TabBasketIcon
            fill={
              activeTab === 'BasketConsumer'
                ? theme.colors.PRIMARY_600
                : theme.colors.GRAY_700
            }
          />,
          'Minha cesta',
          'BasketConsumer',
        )}

        {renderTab(
          <TabFinacial
            fill={
              activeTab === 'DonationConsumer'
                ? theme.colors.PRIMARY_600
                : theme.colors.GRAY_700
            }
          />,
          'Doações',
          'DonationConsumer',
        )}

        {renderTab(
          <TabProfileIcon
            fill={
              activeTab === 'ProfileConsumer'
                ? theme.colors.PRIMARY_600
                : theme.colors.GRAY_700
            }
          />,
          'Perfil',
          'ProfileConsumer',
        )}
      </StyledButtonTabRow>
    </StyledBottomTabContainer>
  );
};

export const StyledBottomTabContainer = styled.SafeAreaView`
  height: 49px;
  background-color: ${({theme}) => theme.colors.GRAY_100};

  box-shadow: 0px -2px 4px rgba(0, 0, 0, 0.12);
`;

export const StyledButtonTabRow = styled.View`
  width: 85%;
  flex-direction: row;
  justify-content: space-between;
  align-self: center;
`;

export const StyledButtonInactive = styled.TouchableOpacity`
  width: 25%;
  align-items: center;
  justify-content: flex-end;
`;

export const StyledTextInactive = styled.Text`
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLEST};
  color: ${({theme}) => theme.colors.GRAY_900};
  margin-top: 5px;
`;

export const StyledTextActive = styled(StyledTextInactive)`
  background-color: ${({theme}) => theme.colors.PRIMARY_100};
  border-radius: 10px;
  padding: 0 10px;
`;
