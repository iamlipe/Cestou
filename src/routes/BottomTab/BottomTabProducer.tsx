import React from 'react';
import Icon from 'react-native-vector-icons/MaterialIcons';
import {useTheme} from 'styled-components/native';
import {BottomTabBarProps} from '@react-navigation/bottom-tabs';
import {
  ParamListBase,
  TabNavigationState,
  useNavigation,
} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {LoggedProducerStackParamList} from '../stacks/LoggedProducerStack';

import {
  StyledBottonTabContainer,
  StyledButtonInactive,
  StyledTextActive,
  StyledTextInactive,
  StyledButtonTabRow,
} from './BottomTabConsumer';

type NavPropsProducer = NativeStackNavigationProp<
  LoggedProducerStackParamList,
  'HomeProducerStack' | 'FinancialProducerStack' | 'ProfileProducer'
>;

interface Props {
  state: TabNavigationState<ParamListBase>;
}

export const ButtonTabProducer: React.FC<Props> = ({state}) => {
  const {navigate} = useNavigation<NavPropsProducer>();
  const activeTab = state.routes[state.index].name;
  const theme = useTheme();

  const renderTab = (
    name: string,
    title: string,
    route: keyof LoggedProducerStackParamList,
    testID: string,
  ) => (
    <StyledButtonInactive onPress={() => navigate(route)}>
      <Icon
        testID={testID}
        name={name}
        size={24}
        color={
          activeTab === route ? theme.colors.PRIMARY_600 : theme.colors.GRAY_700
        }
      />
      {activeTab === route ? (
        <StyledTextActive>{title}</StyledTextActive>
      ) : (
        <StyledTextInactive>{title}</StyledTextInactive>
      )}
    </StyledButtonInactive>
  );

  return (
    <StyledBottonTabContainer>
      <StyledButtonTabRow>
        {renderTab('home', 'In√≠cio', 'HomeProducerStack', 'icon-tab-home')}
        {renderTab(
          'attach-money',
          'Financeiro',
          'FinancialProducerStack',
          'icon-tab-financial',
        )}
        {renderTab('person', 'Perfil', 'ProfileProducer', 'icon-tab-profile')}
      </StyledButtonTabRow>
    </StyledBottonTabContainer>
  );
};
