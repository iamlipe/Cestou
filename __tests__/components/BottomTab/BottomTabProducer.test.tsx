import React from 'react';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@/config/functions/testing';
import {mockStackTabNavigatorProducer} from '@__mocks__/mockStackTabNavigator';

import {ButtonTabProducer} from '@/routes/BottomTab/BottomTabProducer';
import {fireEvent} from '@testing-library/react-native';

describe('ButtonTabProducer', () => {
  test('should render correctly', () => {
    renderWithThemeProvider(
      <ButtonTabProducer state={mockStackTabNavigatorProducer} />,
    );
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(
      <ButtonTabProducer state={mockStackTabNavigatorProducer} />,
    );
  });

  test('should start select tab home', () => {
    const buttonTab = renderWithThemeProvider(
      <ButtonTabProducer state={mockStackTabNavigatorProducer} />,
    );

    const tabIconHome = buttonTab.getByTestId('icon-tab-home');

    console.log(tabIconHome.props.style[0].color);

    expect(tabIconHome.props.style[0].color).toBe('#00A857');
  });

  test('should select tab financial', () => {
    const buttonTab = renderWithThemeProvider(
      <ButtonTabProducer
        state={{...mockStackTabNavigatorProducer, index: 1}}
      />,
    );

    const tabIconFinancial = buttonTab.getByTestId('icon-tab-financial');

    fireEvent.press(tabIconFinancial);

    const newTabIconFinancial = buttonTab.getByTestId('icon-tab-financial');

    expect(newTabIconFinancial.props.style[0].color).toBe('#00A857');
  });

  test('should select tab profile', () => {
    const buttonTab = renderWithThemeProvider(
      <ButtonTabProducer
        state={{...mockStackTabNavigatorProducer, index: 2}}
      />,
    );

    const tabIconProfile = buttonTab.getByTestId('icon-tab-profile');

    fireEvent.press(tabIconProfile);

    const newTabIconProfile = buttonTab.getByTestId('icon-tab-profile');

    expect(newTabIconProfile.props.style[0].color).toBe('#00A857');
  });
});
