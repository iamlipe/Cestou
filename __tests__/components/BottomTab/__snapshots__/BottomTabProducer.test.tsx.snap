// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonTabProducer should matches snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "backgroundColor": "#F2F2F8",
        "height": 49,
        "shadowColor": "rgba(0, 0, 0, 0.12)",
        "shadowOffset": Object {
          "height": -2,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 4,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": "85%",
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "flex-end",
          "opacity": 1,
          "width": "25%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#00A857",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
        testID="icon-tab-home"
      >
        
      </Text>
      <Text
        style={
          Array [
            Object {
              "backgroundColor": "#C1E9CF",
              "borderRadius": 10,
              "color": "#1A1A1A",
              "fontFamily": "SourceSansPro-Regular",
              "fontSize": 24,
              "marginTop": 5,
              "paddingBottom": 0,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 0,
            },
          ]
        }
      >
        Início
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "flex-end",
          "opacity": 1,
          "width": "25%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#3A3A3A",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
        testID="icon-tab-financial"
      >
        
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#1A1A1A",
              "fontFamily": "SourceSansPro-Regular",
              "fontSize": 24,
              "marginTop": 5,
            },
          ]
        }
      >
        Financeiro
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "flex-end",
          "opacity": 1,
          "width": "25%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#3A3A3A",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
        testID="icon-tab-profile"
      >
        
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#1A1A1A",
              "fontFamily": "SourceSansPro-Regular",
              "fontSize": 24,
              "marginTop": 5,
            },
          ]
        }
      >
        Perfil
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
